File: patterns/architectural/microservices.js
/**
 * Microservices Pattern - JavaScript
 * 
 * Architecture where applications are built as a collection of small,
 * independent services that communicate over well-defined APIs.
 */

// Service 1: User Service
class UserService {
    constructor() {
        this.users = [];
    }
    
    createUser(userData) {
        const user = {
            id: Date.now(),
            ...userData,
            createdAt: new Date()
        };
        this.users.push(user);
        return user;
    }
    
    getUser(id) {
        return this.users.find(u => u.id === id);
    }
}

// Service 2: Order Service
class OrderService {
    constructor(userService) {
        this.userService = userService;
        this.orders = [];
    }
    
    createOrder(userId, items) {
        const user = this.userService.getUser(userId);
        if (!user) {
            throw new Error('User not found');
        }
        
        const order = {
            id: Date.now(),
            userId,
            items,
            total: items.reduce((sum, item) => sum + item.price, 0),
            createdAt: new Date()
        };
        this.orders.push(order);
        return order;
    }
    
    getOrder(id) {
        return this.orders.find(o => o.id === id);
    }
}

// API Gateway (simplified)
class APIGateway {
    constructor(userService, orderService) {
        this.userService = userService;
        this.orderService = orderService;
    }
    
    handleRequest(method, path, data) {
        if (path.startsWith('/users')) {
            return this.handleUserRequest(method, path, data);
        } else if (path.startsWith('/orders')) {
            return this.handleOrderRequest(method, path, data);
        }
        return { error: 'Not found' };
    }
    
    handleUserRequest(method, path, data) {
        if (method === 'POST' && path === '/users') {
            return this.userService.createUser(data);
        } else if (method === 'GET' && path.startsWith('/users/')) {
            const id = parseInt(path.split('/')[2]);
            return this.userService.getUser(id);
        }
        return { error: 'Not found' };
    }
    
    handleOrderRequest(method, path, data) {
        if (method === 'POST' && path === '/orders') {
            return this.orderService.createOrder(data.userId, data.items);
        } else if (method === 'GET' && path.startsWith('/orders/')) {
            const id = parseInt(path.split('/')[2]);
            return this.orderService.getOrder(id);
        }
        return { error: 'Not found' };
    }
}

// Example usage
console.log('=== Microservices Pattern ===\n');

const userService = new UserService();
const orderService = new OrderService(userService);
const gateway = new APIGateway(userService, orderService);

// Create user via gateway
const user = gateway.handleRequest('POST', '/users', {
    name: 'John Doe',
    email: 'john@example.com'
});
console.log('Created user:', user);

// Create order via gateway
const order = gateway.handleRequest('POST', '/orders', {
    userId: user.id,
    items: [
        { name: 'Product 1', price: 10 },
        { name: 'Product 2', price: 20 }
    ]
});
console.log('Created order:', order);

module.exports = { UserService, OrderService, APIGateway };

File: patterns/architectural/microservices.py
"""
Microservices Pattern - Python

Architecture where applications are built as a collection of small,
independent services that communicate over well-defined APIs.
"""


# Service 1: User Service
class UserService:
    def __init__(self):
        self.users = []
    
    def create_user(self, user_data):
        import time
        user = {
            'id': int(time.time()),
            **user_data,
            'created_at': time.time()
        }
        self.users.append(user)
        return user
    
    def get_user(self, id):
        return next((u for u in self.users if u['id'] == id), None)


# Service 2: Order Service
class OrderService:
    def __init__(self, user_service):
        self.user_service = user_service
        self.orders = []
    
    def create_order(self, user_id, items):
        user = self.user_service.get_user(user_id)
        if not user:
            raise ValueError("User not found")
        
        import time
        order = {
            'id': int(time.time()),
            'user_id': user_id,
            'items': items,
            'total': sum(item['price'] for item in items),
            'created_at': time.time()
        }
        self.orders.append(order)
        return order
    
    def get_order(self, id):
        return next((o for o in self.orders if o['id'] == id), None)


# API Gateway (simplified)
class APIGateway:
    def __init__(self, user_service, order_service):
        self.user_service = user_service
        self.order_service = order_service
    
    def handle_request(self, method, path, data=None):
        if path.startswith('/users'):
            return self.handle_user_request(method, path, data)
        elif path.startswith('/orders'):
            return self.handle_order_request(method, path, data)
        return {'error': 'Not found'}
    
    def handle_user_request(self, method, path, data):
        if method == 'POST' and path == '/users':
            return self.user_service.create_user(data)
        elif method == 'GET' and path.startswith('/users/'):
            id = int(path.split('/')[2])
            return self.user_service.get_user(id)
        return {'error': 'Not found'}
    
    def handle_order_request(self, method, path, data):
        if method == 'POST' and path == '/orders':
            return self.order_service.create_order(data['user_id'], data['items'])
        elif method == 'GET' and path.startswith('/orders/'):
            id = int(path.split('/')[2])
            return self.order_service.get_order(id)
        return {'error': 'Not found'}


# Example usage
if __name__ == "__main__":
    print("=== Microservices Pattern ===\n")
    
    user_service = UserService()
    order_service = OrderService(user_service)
    gateway = APIGateway(user_service, order_service)
    
    # Create user via gateway
    user = gateway.handle_request('POST', '/users', {
        'name': 'John Doe',
        'email': 'john@example.com'
    })
    print("Created user:", user)
    
    # Create order via gateway
    order = gateway.handle_request('POST', '/orders', {
        'user_id': user['id'],
        'items': [
            {'name': 'Product 1', 'price': 10},
            {'name': 'Product 2', 'price': 20}
        ]
    })
    print("Created order:", order)

